### PR TITLE
Add example standalone spark with S3 storage

### PR DESCRIPTION
**Pull request by fhoering**

### CODE REVIEW

**Comment by t-henri** [comment]
I would make a separate PR for the github actions configuration, here neither travis or github actions ran

File: examples/spark-with-S3/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# docker file/docker-compose derived from https://github.com/gettyimages/docker-spark
+# https://github.com/gettyimages/docker-spark/blob/master/LICENSE
+FROM debian:stretch
+
+RUN apt-get update
+#    && apt-get install -y curl unzip \
+#    && apt-get install -y python-pkg-resources python3.6 python3.6-venv python3-venv
```
Comment (t-henri) [inline examples/spark-with-S3/Dockerfile:7]: If they are commented can we remove these ?

File: examples/spark-with-S3/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  | gunzip \
+  | tar x -C /usr/ \
+ && mv /usr/$SPARK_PACKAGE $SPARK_HOME \
+ && chown -R root:root $SPARK_HOME
+
+WORKDIR /cluster-pack
+
+
+
+
```
Comment (t-henri) [inline examples/spark-with-S3/Dockerfile:55]: We can keep a single blank line at the end

File: cluster_pack/spark/spark_config_builder.py [hunk 3]
```diff
@@ -20,20 +28,48 @@ def add_packaged_environment(ssb: SparkSession.Builder, archive: str):
...
     for requirement_dir in packaging.get_editable_requirements().values():
         py_archive = packaging.zip_path(requirement_dir)
         _add_archive(ssb, py_archive)
 
 
+def add_s3_params(ssb: SparkSession.Builder,  fs_args: Dict[str, Any] = {}):
+    ssb.config("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
+    ssb.config("spark.hadoop.fs.s3a.path.style.access", "true")
+    if "client_kwargs" in fs_args:
+        ssb.config("spark.hadoop.fs.s3a.endpoint", fs_args["client_kwargs"]["endpoint_url"])
```
Comment (t-henri) [inline cluster_pack/spark/spark_config_builder.py:46]: Can we let the user specify ACCESS_KEY/SECRET_KEY here ?

File: cluster_pack/filesystem.py [hunk 4]
```diff
@@ -8,6 +8,12 @@
 
 try:
     from s3fs import S3FileSystem
+
+    def _makedirs(self, path, exist_ok=False):
+        bucket, _, _ = self.split_path(path)
+        if not self.exists(bucket):
+            self.mkdir(bucket)
+    S3FileSystem.makedirs = _makedirs
```
Comment (t-henri) [inline cluster_pack/filesystem.py:16]: Why do you need to overwite S3 makedirs function if you never use it ?

File: cluster_pack/spark/spark_config_builder.py [hunk 5]
```diff
@@ -1,15 +1,23 @@
...
 import os
-
+import logging
 import pyspark
 from pyspark.sql import SparkSession
 
+try:
+    from s3fs import S3FileSystem
+except ModuleNotFoundError:
+    pass
```
Comment (t-henri) [inline cluster_pack/spark/spark_config_builder.py:10]: Why do you need to import S3FileSystem if you never use it ?

File: examples/spark-with-S3/Dockerfile (hunk 1 above)
Comment (fhoering) [reply examples/spark-with-S3/Dockerfile]: Done

File: examples/spark-with-S3/Dockerfile (hunk 2 above)
Comment (fhoering) [reply examples/spark-with-S3/Dockerfile]: Done

File: cluster_pack/spark/spark_config_builder.py (hunk 3 above)
Comment (fhoering) [reply cluster_pack/spark/spark_config_builder.py]: Done

File: cluster_pack/spark/spark_config_builder.py (hunk 5 above)
Comment (fhoering) [reply cluster_pack/spark/spark_config_builder.py]: Removed

File: cluster_pack/filesystem.py (hunk 4 above)
Comment (fhoering) [reply cluster_pack/filesystem.py:18]: - arrow calls mkdirs https://github.com/apache/arrow/blob/master/python/pyarrow/filesystem.py#L336
- in S3FS mkdirs is an alias to makedirs which is implemented as no-op (def makedirs() pass:)

I fixed it with https://github.com/dask/s3fs/pull/331 but we need to wait for a release.

Added a comment in the code.

File: cluster_pack/spark/spark_config_builder.py [hunk 6]
```diff
@@ -20,20 +23,52 @@ def add_packaged_environment(ssb: SparkSession.Builder, archive: str):
...
         _add_archive(ssb, py_archive)
 
 
+def add_s3_params(ssb: SparkSession.Builder,  fs_args: Dict[str, Any] = {}):
+    ssb.config("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
+    ssb.config("spark.hadoop.fs.s3a.path.style.access", "true")
+    if "key" in fs_args:
+        ssb.config("spark.hadoop.fs.s3a.access.key", fs_args["key"])
+    if "secret" in fs_args:
+        ssb.config("spark.hadoop.fs.s3a.secret.key", fs_args["secret"])
```
Comment (t-henri) [inline cluster_pack/spark/spark_config_builder.py:43]: I would have named the args 'access_key' and 'secret_key'

Comment (fhoering) [reply cluster_pack/spark/spark_config_builder.py:43]: I reused the args of:
https://s3fs.readthedocs.io/en/latest/api.html#s3fs.core.S3FileSystem

such that we can just do:
```
s3fs = new S3FileSystem(**fs_args)
```

Comment (fhoering) [reply cluster_pack/spark/spark_config_builder.py:43]: also applies to this (client_kwargs : dict of parameters for the botocore client)
```
 if "client_kwargs" in fs_args:
        ssb.config("spark.hadoop.fs.s3a.endpoint", fs_args["client_kwargs"]["endpoint_url"])
```
we could directly pass the url also

Comment (fhoering) [reply cluster_pack/spark/spark_config_builder.py:43]: Personally I prefer just reusing the args from S3FileSystem but it I agree it is debatable. What happens if we want to use the new S3 client of pyarrow (with other names for the args, but it seems not mature yet)

Comment (t-henri) [reply cluster_pack/spark/spark_config_builder.py:43]: Ok it's fine for me, I didn't check S3FileSystem documentation before my comment

**Pull request merged**